### ISSUE TITLE
Use file or AudioSource as input?

### ISSUE THREAD

**Issue by ZephyrRaine** [keijiro/Lasp#2]
Hello keijiro, thanks for that and all of your work, you're really bringing great and useful features to Unity 🥇   
  
I was wondering if there was any way to use a file as input for Lasp, or if it could be done easily by modifying your source code?
It works fine when setting Stereo Mix as a recording device, but it probably won't be enabled for the random user.
Not being that much of a savvy guy, I was wondering if you'd have any clue on that for me to get started! 
  
Thanks in advance!

**Comment by keijiro** [comment]
> I was wondering if there was any way to use a file as input for Lasp, or if it could be done easily by modifying your source code?

No, but I have a plan to add "feedback mode" that allows using Unity's Audio Mixer as an input to LASP.

**Comment by ZephyrRaine** [comment]
Thanks for your answer, looking forward to it. 
If there's any way we can help let us know :)

**Comment by AnyMotion** [comment]
Feedback mode sounds exactly like the function I'm looking for. Reading individual channel/group levels and give rms / peak values from C++ function. If that's a possibility, than I'm a happy camper. :)

If only Unity would also allow more script access to the mixer, like creating channels, groups and exposed variables on the fly (and not naming them myexposedreference every time :( ) and give us access to the peak/rms meters already build in Unity. I tried some native audio things, but with no experience in C++ this isn't for me.

**Comment by keijiro** [comment]
I implemented the loopback feature as a special build.

https://github.com/keijiro/Lasp/tree/loopback

I won't merge it to the master branch for a while because it hasn't been tested enough.

**Comment by AnyMotion** [comment]
Cool Thx. Will test this soon and let you know the results. Keep up the good work, unity should promote you to chief audio!

Any idea if there will be a possibility to add exposed parameters / effects etc through c# script to the unity mixer anytime soon. It would really optimize my project vr sound engineer and make a lot more possible than the current limited way of using exposed values. (I need about 2000 of them and the dorpdown list in unity mixer is not an easy workflow and quite buggy as some exp values (specially the send effect) tend to lose their reference.)

Regards,
Marald Bes
AnyMotion.nl

Verstuurd vanaf mijn iPhone

**Comment by AnyMotion** [comment]
Hi Kijiro,

It took me a while, but was able to test the Last loopback in Unity 2017.2 today.
It seems to work fine (lower cpu usage than reading audiofilter output), but I ran into some limitations.

As the implementation is done using audioinput, it is general for the entire project.
I wanted to receive RMS / Peakvalues of each separate channel where the Lasp loopback plugin is inserted and enabled.
Is there a way to do this? Is it even possible to use multiple LASP plugins simultaniously?

In the screenshot you can see the current LASP loop effect on the meters of the mixing console. All the meters seem to be using the same Lasp plugin. The only difference you see in levels is because of different filter types which is configured per channel.



Kind regards,
Marald Bes
www.AnyMotion.nl

**Comment by keijiro** [comment]
> Is there a way to do this? Is it even possible to use multiple LASP plugins simultaniously?

No. LASP doesn't support multiple channels at the moment. I think it should be supported though.

**Comment by stefang** [comment]
Wondering if this is in the current new build? Guessing not by the fact it's still an open ticket?

Excited to try this out if it is... Would it be safe to assume that the combo of the above mentioned loopback mode, an audioSource and the Timeline would solve the issue of offline rendering visuals using the Unity Recorder and keeping the rendered files in sync with an source audio file? The not-realtime rendering of an image sequence means live audio input isn't an option.

**Comment by keijiro** [comment]
@stefang No, it isn't in the current build. My current conclusion is that I can't afford enough time to support it in a proper way. Please understand that my capacity it quite limited.

**Comment by stefang** [comment]
Of course, I was just checking really! Grateful you make any of this available :-)

**Comment by Saggex** [comment]
Hi Keijiro,

Is there a plan to upgrade the Loopback Branch to the V2 of LASP?

Thank you and BR :)

**Comment by Gabana** [comment]
Hi Keijiro,

Any chance to use Unity's Audio Source as a source for LASP in the latest updates you did recently (mid 2020)?? I see you made some attempts but the Loopback branch doesn't work anymore due to a problem in Lasp.bundle (in 2019.4), i wonder if you have a different solution. 

I saw one of commentend a possible solution, where i should "create a fake input device and somehow route the sound trough there"... Can you please provide any hint about to do achieve this?

I just want to play a song using LASP to make an audio visualization using VFX Graph...

Any help would be greatly appreciated :)

**Comment by keijiro** [comment]
@Gabana 

> i wonder if you have a different solution.

No. There is no solution.

> I saw one of commentend a possible solution, where i should "create a fake input device and somehow route the sound trough there"... Can you please provide any hint about to do achieve this?

Virtual audio device like this: https://vac.muzychenko.net/en/

**Comment by Nolram12345** [comment]
Still no further update on this ?

**Comment by Saggex** [comment]
> Still no further update on this ?


I just implemented a solution a few minutes ago. 
It does not yet support the AudioSpectrumAnalyzer Component, but It works with the AudioLevelTracker.
It's for now on my google drive, until I find the time to create a fork of the repo, or until I find the time, to create a branch here, or something.
You have to remove LASP from your packages and put the contents into your Assets folder.

https://drive.google.com/file/d/1cwjt9mjDGC5CS9wxRHGeQF4q-yBo15d_/view?usp=sharing

**Comment by Thaina** [comment]
I can easily use raw data from AudioSource and convert to texture easily with this

```C#
            if(audioSource && audioSource.isPlaying)
            {
                if(audioData?.Length != _analyzer.resolution)
                    audioData = new float[_analyzer.resolution];

                audioSource.GetOutputData(audioData,0);
                
                EnsureSize(ref cacheAudioData,audioData.Length);

                cacheAudioData.CopyFrom(audioData);

                if(fft?.Width != audioData.Length * 2)
                {
                    fft?.Dispose();
                    fft = new FftBuffer(audioData.Length * 2);
                }

                fft.Push(cacheAudioData);
                fft.Analyze(-_analyzer.currentGain - _analyzer.dynamicRange, -_analyzer.currentGain);
                
                if(logScaler == null)
                    logScaler = new LogScaler();

                data = logScaler.Resample(fft.Spectrum);
            }
            else data = _logScale ? _analyzer.logSpectrumArray : _analyzer.spectrumArray;
```

It require us to made some internal class become public like this https://github.com/keijiro/Lasp/pull/65

I would also suggest disable autogain with audiosource

**Comment by anttimo** [comment]
Hi! Since I got this working with my RME audio interface (UCX in my case) I thought I mention it here as it might be useful to others. 

**With RME audio interface's TotalMix software:**
- Turn on Loopback on for one of your hardware outputs that is not in use. In my case "AN 1/2".

<img width="87" alt="Image" src="https://github.com/user-attachments/assets/e7d5207f-193a-443e-b6ee-a19858761b07" />

**In Unity:**
- For example in your Audio Level Tracker, turn off the default device and then select the Device ID matching the hardware output you turned on the loopback for.
-> Now it should work! If not make sure you turn up the volume in the TotalMix for the selected hardware output.

Here's RME documentation about the Loopback feature https://rme-audio.de/totalmix-fx-record-all-sounds.html.

I assume this should work with any audio interface that supports loopback. Maybe this was obvious info for some but I'm quite new to all this so I thought I mention this as somebody might find this info useful :)
